src/utility.h

```diff
@@ -0,0 +1,331 @@
+        // this non-copyable type in containers.
+        m_fd = other.get();
+        other.invalidate();
+    }
+
+    ~FileDesc()
+    {
+        if (valid())
+        {
+            close();
+        }
+    }
+
+    FileDesc(const FileDesc &other) = delete;
+    FileDesc& operator=(const FileDesc &other) = delete;
+
+    int get() const { return m_fd; }
+
+    void set(int fd)
+    {
+        if (valid())
+        {
+            close();
+        }
+
+        m_fd = fd;
+    }
+
+    void steal(FileDesc &other)
+    {
+        set(other.get());
+        other.invalidate();
+    }
+
+    bool valid() const { return m_fd != -1; }
+    bool invalid() const { return !valid(); }
+    void invalidate() { m_fd = -1; }
+
+    //! explicitly close and invalidate() the currently stored file descriptor
+    void close();
+
+protected:
+
+    int m_fd = -1;
+};
+
+//! C++ wrapper around the POSIX struct stat
+class FileStatus :
+    public ::stat
+{
+public:
+
+    bool isLink() const { return S_ISLNK(this->st_mode); }
+    bool isRegular() const { return S_ISREG(this->st_mode); }
+    bool isDirectory() const { return S_ISDIR(this->st_mode); }
+
+    //! returns the file mode bits portion consisting of the permission bits
+    //! plus any special bits like setXid but not the file type bits
+    auto getModeBits() const { return this->st_mode & 07777; }
+
+    bool matchesOwnership(const uid_t uid, const gid_t gid) const
+    {
+        return this->st_uid == uid && this->st_gid == gid;
+    }
+
+    //! returns whether this file status and the other file status refer to
+    //! the same file object (based on device and inode identification)
+    bool sameObject(const struct ::stat &other) const
+    {
+        return this->st_dev == other.st_dev && this->st_ino == other.st_ino;
+    }
+
+    bool isWorldWritable() const
+    {
+        return (this->st_mode & S_IWOTH) != 0;
+    }
+
+    void copy(const struct stat &other)
+    {
+        std::memcpy(this, &other, sizeof(*this));
+    }
+
+    bool fstat(int fd)
+    {
+        return ::fstat(fd, this) == 0;
+    }
+};
+
+//! a wrapper around the native cap_t type to ease memory management
+class FileCapabilities
+{
+public:
+
+    explicit FileCapabilities() {}
+
+    ~FileCapabilities();
+
+    FileCapabilities(FileCapabilities &&other)
+    {
+        // steal the rvalue's caps so we take over ownership, while
+        // the other doesn't free them during destruction. This allows to keep
+        // this non-copyable type in containers.
+        m_caps = other.m_caps;
+        other.invalidate();
+    }
+
+    FileCapabilities(const FileCapabilities &other) = delete;
+    FileCapabilities& operator=(const FileCapabilities &other) = delete;
+
+    bool operator==(const FileCapabilities &other) const;
+    bool operator!=(const FileCapabilities &other) const
+    {
+        return !(*this == other);
+    }
+
+    bool valid() const { return m_caps != nullptr; }
```

Comment:
I think it's open for debate. As I understand it the cap interface doesn't differentiate between an error and "no data available", except via the `errno` returned. So it could be both, either `!valid()`, because an error occured, or `empty()` because there's no data available.

Reply:
True. `cap_get_file/cap_get_fd` errors out with `ENODATA` when there are no capabilities (undocumented, unfortunately). `cap_from_text` does return a non-NULL object when given an empty string on the other hand. So the underlying library is inconsistent :(

The way the program deals with this currently (e.g. `ProfileEntry::hasCaps()`, `ctx.caps` only being set when an extra profile entry exists), is more like 'empty' than 'invalid' though.

Reply:
I'm still not quite sure about this notion. Of course we can change the name to `empty()` but I have the feeling that then it's wrong just the other way around. The current application logic simply is smudgy. It assumes the meaning to be `empty()` which is true in most cases. But if an actual error occurs then this assumption will no longer hold. And then some of the descisions made in the code will probably also be unwise.

Furthermore the current usage of the `ProfileEntry.caps` attribute is unclean. In my view it should only represent what is found in the profile data on disk and never the *actual* capabilities of the target files. This should be separate. But this was too hot an iron for me to touch for the refactoring. I'd like to postpone it.

Reply:
Indeed, the refactoring is just surfacing existing the problem that the program is not differentiating the different meanings of `NULL`.

I can understand if you don't want to fix that completely now, but then please document this issue at the usages of `valid` that actually treat it as `empty`.